### DISCUSSION TITLE
'workspace' endgame

### DISCUSSION THREAD

**Discussion by majidaldo** [astral-sh/rye#314] [General]
what's the vision for creating 'local dependencies'? would it be possible to create project deps from workspace dep graphs  ?

here's where i'm coming from https://github.com/pnnl/hydraconda/tree/master/project. i've used conda and the ability to compose ugly jinja templated yaml files to create simultaneous execution environments mainly for data science scenarios where i don't care about packaging too much. but it would be nice to converge on some of the ideas expressed in rye.

i've contemplated switching to nix for my programming-language agnostic needs but it's difficult to use!

**Comment by bluss** [comment]
I would like to see local dependencies work with relative paths primarily. Both of these:

A. depencency = "./subdirectory"
B. dependency = "../sibling_directory"

Because, sibling directories is one way to handle monorepo style projects. Especially those where lots of disparate projects all live in the same repo.

Current workspaces are a way to have local dependencies without declaring them, so they work like an alternative to (A) - all packages in a workspace are just installed together, so they can be used together without even declaring the dependency. Ideally it should be declared, shouldn't it?  Workspaces can't support the scenario (B) - not the most general case of it, so that makes it seem like relative path dependencies are more important than workspaces(?)

The ideal I see is something I learned from Rust (too), you declare a dependency like  foo = { version = "1", path = "./foo" } and it's a path dependency in-tree, but once turned into a wheel it's just a named dependency (dependency on "foo"), which will work fine if both packages are built and installed together.

**Reply by cnpryer** [reply]
+1 for cargo-managed workspaces. So far I've enjoyed using workspaces in rust.

**Reply by bluss** [reply]
We need to map out some similarities and differences between cargo workspaces and what we can do in python I think. Rust is compiled, so workspaces are partly just like an optimization - pooling intermediate builds and dependencies together, but it *would work* without workspaces too, just copy compiled artifacts together correctly like cargo would normally do.

In Python - what *should* it mean to have a workspace? What's the benefit and drawback? Is it this?:

1. Common lockfile
2. Common virtualenv

Are both of these obvious, or can they be questioned? And is something missing from the list?

**Reply by cnpryer** [reply]
In an ideal world it's a bunch of things to me, but mainly an abstraction for managing my projects together.

> pooling intermediate builds and dependencies together

This. 

> 1. Common lockfile
> 2. Common virtualenv

This and probably 3. Common workflow.

**Reply by majidaldo** [reply]
> I would like to see local dependencies work with relative paths primarily. Both of these:
> 
> A. depencency = "./subdirectory" B. dependency = "../sibling_directory"
> 
it should resemble a build system. i don't think using the filesystem to encode dependency is flexible enough.

**Reply by majidaldo** [reply]
> In an ideal world it's a bunch of things to me, but mainly an abstraction for managing my projects together.
> 
> > pooling intermediate builds and dependencies together
> 
> This.
> 
> > 1. Common lockfile
> > 2. Common virtualenv
> 
> This and probably 3. Common workflow.

i like the abstraction idea.  perhaps a separate project.tomls that somehow connectsto pyproject tomls. again, it would resemble a build system.

**Reply by bluss** [reply]
> it should resemble a build system. i don't think using the filesystem to encode dependency is flexible enough.

I would prefer local information instead of central and global, so that each subpart of the repository is understandable mostly as an independent unit, so anything that fits with this.  Do you have something in mind with resembling a build system? How do you specify dependencies in that case?  The basic unit is a pyproject.toml here, is it? (The unit that can have dependencies and be depended on.)

**Reply by cnpryer** [reply]
> I would prefer local information instead of central and global, so that each subpart of the repository is understandable mostly as an independent unit

When you say central and global, do you mean global to the workspace?

**Reply by bluss** [reply]
To just expand on it, hope it clarifies the point:
- local information: subpart A declares - in its pyproject - it depends on subpart B.
- global information (variant 1): A and B can use each other because they are in the same workspace. They don't declare any dependency on each other.
- global information (variant 2): A depends on B but it's written in a makefile in the repository root

local is good because it makes it much easier to lift one subtree of the whole project from one repository to another. And refactorability should be good.

It's not strictly necessary to use file paths, you can just declare A depends on B by other means, too.

**Reply by cnpryer** [reply]
Ignoring implementation details, why wouldn't local information also work in a workspace that supports a global information variant 2 style option as well? Just like in rust [with cargo](https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html#creating-the-second-package-in-the-workspace) you can declare local information and defer to workspace information if needed.

I think I'm just struggling to understand why global vs local is important for project composability (implementation details?). My thought process is to first think about what experience we want, and then think about the implementation required for that experience (if any).

**Reply by cnpryer** [reply]
I get that the support for this in the ecosystem isn't there, but assuming we want to explore new options is there anything else that keeps this off the table?

**Reply by bluss** [reply]
cargo always uses local information (a crate must define its own dependencies, without that it doesn't compile). This point doesn't need to be exaggerated, I think we're just trying to clarify what we actually mean.

I want to avoid dependency relations that are implicit or never defined. If I want to copy one library from a repo and its dep information is not separately available, that's friction and the kind of problem with composability I mean. If I want to add a new library and have to manually merge together two different build system files, that limits composability. :slightly_smiling_face:

**Comment by majidaldo** [comment]
> > it should resemble a build system. i don't think using the filesystem to encode dependency is flexible enough.
> 
> I would prefer local information instead of central and global, so that each subpart of the repository is understandable mostly as an independent unit, so anything that fits with this. Do you have something in mind with resembling a build system? How do you specify dependencies in that case? The basic unit is a pyproject.toml here, is it? (The unit that can have dependencies and be depended on.)

 i very much appreciate this. but the filesystem encoded dependency is not flexible. such a dependency graph should be generated. (what if something else in the 'root' depends on something you have as a deep 'child'?)
 
 see my project above for how i encoded deps via yamls files... makefile style.

**Reply by bluss** [reply]
Maybe a flatter structure is better, in that case? I'm not sure what's wrong with just giving the path to where the dependency is, even if it's nested.  For the yaml files, I don't know what to look at, I'm sorry, they are not too easy to read without introduction to it :slightly_smiling_face:  I see a list of `dependencies` but that's on pip packages, not one part of the repo depending on a lib in another part of the repo, which I'm looking for.

**Comment by majidaldo** [comment]
> Maybe a flatter structure is better, in that case? I'm not sure what's wrong with just giving the path to where the dependency is, even if it's nested. For the yaml files, I don't know what to look at, I'm sorry, they are not too easy to read without introduction to it 🙂 I see a list of `dependencies` but that's on pip packages, not one part of the repo depending on a lib in another part of the repo, which I'm looking for.

sorry about that. the 'workspace' deps are the 'includes' keys.

**Comment by majidaldo** [comment]
> > Maybe a flatter structure is better, in that case? I'm not sure what's wrong with just giving the path to where the dependency is, even if it's nested. For the yaml files, I don't know what to look at, I'm sorry, they are not too easy to read without introduction to it 🙂 I see a list of `dependencies` but that's on pip packages, not one part of the repo depending on a lib in another part of the repo, which I'm looking for.
> 
> sorry about that. the 'workspace' deps are the 'includes' keys.
..oh and since those under 'dependencies' are conda pkgs, they are not necessarily (python) pip.
